### DISCUSSION TITLE
WebGL library is 2x slower in NextJS compared to other static page frameworks

### DISCUSSION THREAD

**Discussion by ingun37** [vercel/next.js#22286] [Help] [answered]
I make a WebGL library that renders realistic garments.
I build & deploy for two environments: cjs(node) and esm(web).
I benchmark my library by loading some heavy models that require lots of resources and time. The result is that cjs is 2x slower than esm. CJS is ran on NextJS while the other on plain HTML. They both run the same bechmarking js code and on the same browser and environment.

I thought it could be a bundling problem so I tried different tools (Webpack and Rollup)with different configuration but the result was same. Bundling wasn't the problem.

Then I tried running the same benchmark code on Gatsby and the result was that it was as fast as ESM module. So I guess the problem occurs only in NextJS. I guess SSR/Pre-rendering is somehow affecting the runtime performance of my library.

How could a library be only slow on NextJS? How can I fix it? Has there been any runtime performance issue about NextJS?

**Comment by ingun37** [comment]
I found it's not only my library or WebGL codes but also unzipping a file using [JSZip](https://stuk.github.io/jszip/) is extremely slow. I think any cjs module are generally slow in Nextjs. I'll write another discussion with sample app that uses JSZip to demonstrate the problem.

**Comment by ingun37** [comment]
I managed to reproduce in a sample app and issued a bug. https://github.com/vercel/next.js/issues/22372

**Comment by timneutkens** [comment, marked as answer]
Replied here: https://github.com/vercel/next.js/issues/22372#issuecomment-783330758
